packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,208 @@
+  private readonly logger: ILogger;
+  private opts: BackfillSyncOpts;
+  //last fetched slot
+  private lastFetchedSlot: Slot | null = null;
+  //last trusted block
+  private anchorBlock: allForks.SignedBeaconBlock | null = null;
+  private processor: ItTrigger = new ItTrigger();
+  private peers: PeerMap<phase0.Status | null> = new PeerMap();
+
+  constructor(modules: RangeSyncModules, opts?: BackfillSyncOpts) {
+    super();
+    this.chain = modules.chain;
+    this.network = modules.network;
+    this.db = modules.db;
+    this.config = modules.config;
+    this.logger = modules.logger;
+    this.opts = opts ?? {batchSize: 32};
+    this.network.events.on(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.on(NetworkEvent.peerDisconnected, this.removePeer);
+    this.network.getConnectedPeers().forEach((peer) => this.peers.set(peer, null));
+    void this.init();
+  }
+
+  /** Throw / return all AsyncGenerators */
+  close(): void {
+    this.network.events.off(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.off(NetworkEvent.peerDisconnected, this.removePeer);
+    this.processor.end(new ErrorAborted("BackfillSync"));
+  }
+
+  async init(): Promise<void> {
+    const finalizedCheckpoint = this.chain.getFinalizedCheckpoint();
+    this.logger.debug("Initializing backfill sync from finalizedCheckpoint", {
+      checkpoint: ssz.phase0.Checkpoint.toJson(finalizedCheckpoint),
+    });
+    this.anchorBlock = await this.db.blockArchive.getByRoot(finalizedCheckpoint.root);
+    void this.sync();
+    this.processor.trigger();
+  }
+
+  async sync(): Promise<void> {
+    try {
+      for await (const _ of this.processor) {
+        this.logger.debug("Backfill Sync - Peer count", {peerCount: this.peers.size});
+        if (this.peers.size === 0) continue;
+
+        const oldestRequiredEpoch = Math.max(
+          GENESIS_EPOCH,
+          this.chain.clock.currentEpoch - getMinEpochForBlockRequests(this.config)
+        );
+        const oldestSlotRequired = computeStartSlotAtEpoch(oldestRequiredEpoch);
+        this.logger.debug("Backfill sync", {
+          oldestSlotRequired,
+          anchorSlot: this.anchorBlock?.message.slot ?? Infinity,
+        });
+        if (this.anchorBlock) {
+          try {
+            const toSlot = this.lastFetchedSlot ?? this.anchorBlock.message.slot;
+            const fromSlot = Math.max(toSlot - this.opts.batchSize, oldestSlotRequired);
+            if (fromSlot >= toSlot) {
+              this.emit(BackfillSyncEvent.completed, toSlot);
+              this.logger.info("Backfill sync completed");
+              break;
+            }
+            this.logger.debug("BackfillSync - range sync", {
+              fromSlot,
+              toSlot,
+              anchorSlot: this.anchorBlock.message.slot,
+              lastFetched: this.lastFetchedSlot ?? -1,
+            });
+            await this.syncRange(fromSlot, toSlot, this.anchorBlock.message.parentRoot);
```

Comment:
Should we be downloading multiple batches at a time as we do in range sync?

Reply:
Had that discussion with dapplion, came to conclusion that speed is not that relevant to backfill sync